compiler/rustc_typeck/src/check/generator_interior.rs

```diff
@@ -384,7 +386,9 @@ impl<'a, 'tcx> Visitor<'tcx> for InteriorVisitor<'a, 'tcx> {
         // However, in the case of temporary values, we are going to store the value into a
         // temporary on the stack that is live for the current temporary scope and then return a
         // reference to it. That value may be live across the entire temporary scope.
-        let scope = if self.drop_ranges.is_borrowed_temporary(expr) {
+        let scope = if self.drop_ranges.is_borrowed_temporary(expr)
+            || ty.map_or(true, |ty| ty.has_significant_drop(self.fcx.tcx, self.fcx.param_env))
```

Comment:
hmm. this is probably more coarse than necessary - it will also trigger for things like `let x = String::new(); drop(x)`. I don't know a better way to do this - I guess this is what you were talking about in DMs?

> eholk: But I think maybe the conservative thing to do is to count the field access (.nickname) as a borrow of the Config temporary, so we should add it to maybe_borrowed_temporaries. Then I think that should make the whole Config live for the whole temporary scope (until the semicolon). It's more conservative than it needs to be, since .nickname has been dropped but this would still consider it live since it's part of the type of Config, but it should be sound.

Reply:
oh, it's still going to use the temporary scope though - I guess the proper use case is something like
```rust
f(String::new()).await
```
where the future doesn't actually use the string

Reply:
oh lol while debugging this I found that there's another existing bug where we're too coarse: for non-Sync types with trivial drop, we still consider them held across the generator:
```
async fn trivial_drop() {
    // uncomment to make this program compile
    // {
    let x: *const usize = &0;
    // }
    async {}.await;
}

fn assert_send<T: Send>(_: T) {}

fn main() {
    assert_send(trivial_drop());
}
```

Reply:
tried to fix this but got an ICE :( I think I will put this off until later
```diff
diff --git a/compiler/rustc_typeck/src/check/generator_interior.rs b/compiler/rustc_typeck/src/check/generator_interior.rs
index 946bfa52d37..5caa1b2ded7 100644
--- a/compiler/rustc_typeck/src/check/generator_interior.rs
+++ b/compiler/rustc_typeck/src/check/generator_interior.rs
@@ -387,7 +387,10 @@ fn visit_expr(&mut self, expr: &'tcx Expr<'tcx>) {
         // temporary on the stack that is live for the current temporary scope and then return a
         // reference to it. That value may be live across the entire temporary scope.
         let scope = if self.drop_ranges.is_borrowed_temporary(expr)
-            || ty.map_or(true, |ty| ty.has_significant_drop(self.fcx.tcx, self.fcx.param_env))
+            || ty.map_or(true, |ty| {
+                !ty.needs_drop(self.fcx.tcx, self.fcx.param_env)
+                || ty.has_significant_drop(self.fcx.tcx, self.fcx.param_env)
+            })
         {
             self.rvalue_scopes.temporary_scope(self.region_scope_tree, expr.hir_id.local_id)
         } else {
```
```
thread 'rustc' panicked at 'index out of bounds: the len is 0 but the index is 13', /home/jnelson/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/ena-0.14.0/src/snapshot_vec.rs:199:10
  21:     0x7f12f71e004b - ena::unify::UnificationTable<S>::value::h616d0c0c6ae4057b
                               at /home/jnelson/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/ena-0.14.0/
src/unify/mod.rs:347:10
  22:     0x7f12f71e004b - ena::unify::UnificationTable<S>::inlined_get_root_key::hea2bf8b4576b5472
                               at /home/jnelson/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/ena-0.14.0/
src/unify/mod.rs:362:19
  23:     0x7f12f71e004b - ena::unify::UnificationTable<S>::inlined_probe_value::hc1dc88e31da4bcc9
                               at /home/jnelson/.local/lib/cargo/registry/src/github.com-1ecc6299db9ec823/ena-0.14.0/
src/unify/mod.rs:566:18
  24:     0x7f12f71e004b - rustc_infer::infer::type_variable::TypeVariableTable::inlined_probe::h29b4e108ffacbc08
                               at /home/jnelson/rust-lang/rust/compiler/rustc_infer/src/infer/type_variable.rs:312:9
  25:     0x7f12f71e004b - rustc_infer::infer::type_variable::TypeVariableTable::probe::h91a8d70bb6dc3e56
                               at /home/jnelson/rust-lang/rust/compiler/rustc_infer/src/infer/type_variable.rs:306:9
  26:     0x7f12f717fe34 - rustc_infer::infer::InferCtxt::probe_ty_var::hefc9501f755a1a65
                               at /home/jnelson/rust-lang/rust/compiler/rustc_infer/src/infer/mod.rs:1388:15
  27:     0x7f12f71d7089 - <rustc_infer::infer::canonical::canonicalizer::Canonicalizer as rustc_middle::ty::fold::Ty
peFolder>::fold_ty::h0ebb9c6ed6bc4918
                               at /home/jnelson/rust-lang/rust/compiler/rustc_infer/src/infer/canonical/canonicalizer
.rs:394:23
...

query stack during panic:
#0 [is_copy_raw] computing whether `[static generator@src/test/ui/async-await/non-trivial-drop-unused.rs:11:11: 11:13
]` is `Copy`
#1 [needs_drop_raw] computing whether `[static generator@src/test/ui/async-await/non-trivial-drop-unused.rs:11:11: 11
:13]` needs drop
#2 [typeck] type-checking `trivial_drop`
#3 [mir_built] building MIR for `trivial_drop`
#4 [unsafety_check_result] unsafety-checking `trivial_drop`
#5 [mir_const] processing MIR for `trivial_drop`
#6 [mir_promoted] processing `trivial_drop`
#7 [mir_borrowck] borrow-checking `trivial_drop`
#8 [type_of] computing type of `trivial_drop::{opaque#0}`
#9 [check_mod_item_types] checking item types in top-level module
#10 [analysis] running analysis passes on this crate
end of query stack
```

Reply:
> I guess the proper test case is something like

looks like this already errors today:
```
$  cat src/test/ui/async-await/non-trivial-drop-unused.rs
use std::rc::Rc;

async fn ignore_param(x: Rc<i32>) {
    drop(x);
    async {}.await;
}

fn assert_send<T: Send>(_: T) {}

fn main() {
    assert_send(ignore_param(Rc::new(0)));
}
$ rustc +nightly src/test/ui/async-await/non-trivial-drop-unused.rs --edition 2018 -Zdrop-tracking 
error: future cannot be sent between threads safely
  --> src/test/ui/async-await/non-trivial-drop-unused.rs:19:17
   |
19 |     assert_send(ignore_param(Rc::new(0)));
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^ future returned by `ignore_param` is not `Send`
   |
   = help: within `impl Future<Output = ()>`, the trait `Send` is not implemented for `Rc<i32>`
note: captured value is not `Send`
  --> src/test/ui/async-await/non-trivial-drop-unused.rs:6:23
   |
6  | async fn ignore_param(x: Rc<i32>) {
   |                       ^ has type `Rc<i32>` which is not `Send`
note: required by a bound in `assert_send`
  --> src/test/ui/async-await/non-trivial-drop-unused.rs:16:19
   |
16 | fn assert_send<T: Send>(_: T) {}
   |                   ^^^^ required by this bound in `assert_send`
```
that *also* seems wrong, for some reason it considers the param held for the whole async function even if you drop it earlier

Reply:
just fyi that neither of those (nor most of the query-like functions on Ty) account for type variables, so that's why it's panicking i think

Reply:
wait what? why do they take a ParamEnv if they don't account for type variables?

Reply:
that's because we still need to normalize projections -- also, reminder that i said type variables, not type parameters. type variables don't have anything to do with param envs. 

note a similar difference between Infcx::type_is_copy_modulo_regions and Ty::is_copy_modulo_regions.

Reply:
Type (inference) variables are associated to a given InferCtxt, so unless the function is associated with one of those, it's not typically safe to pass a Ty to it unless you know that we're past the type-checking phase or we're outside of a function body.

Reply:
thanks, using `!self.fcx.inh.infcx.resolve_vars_if_possible(ty).needs_drop(self.fcx.tcx, self.fcx.param_env)` fixed the ICE (but didn't allow the code to compile for reasons I don't understand - going to stop harping on it in this PR since it's not related to the current fix)

Reply:
nit: fcx derefs to infcx so you should just be able to do fcx.resolve_vars_if_possible

also that'll still ice if you have any region variables, so you might need to pass it through erase_regions for types that have lifetimes and also significant drops

Reply:
Ok, figured this out: https://github.com/rust-lang/rust/pull/99105